README.md
# animated_fab_menu

Simple animated floating action button that transform to menu (or any other widget)

<img src="result.gif" width="200px">

## Getting Started

```
    Scaffold(
      appBar: AppBar(
        title: Text('TestView'),
        centerTitle: true,
      ),
      body: SafeArea(
          child: Stack(
        children: [
          Center(
            child: Text('Hello World!'),
          ),
          AddFab(
            color: Colors.red,                          //optional
            colorTo: Colors.white,                      //optional
            expandHeight: 400,                          //optional
            expandWidth: 150,                           //optional
            btnSize: 70,                                //optional
            icon: Icon(                                 //required
              Icons.add,
              color: Colors.white,
              size: 40,
            ),
            menu: Container(                            //required
              padding: EdgeInsets.symmetric(vertical: 20),
              decoration: BoxDecoration(
                color: Colors.white,
                borderRadius: BorderRadius.circular(20),
              ),
              child: ListView.builder(
                itemCount: 6,
                itemBuilder: (context, i) => Container(
                  padding: EdgeInsets.symmetric(vertical: 5),
                  alignment: Alignment.center,
                  width: double.infinity,
                  decoration: BoxDecoration(color: Colors.white),
                  child: Text(
                    '$i',
                    style: TextStyle(fontSize: 25),
                  ),
                ),
              ),
            ),
          ),
        ],
      )),
    );
```


CHANGELOG.md
## [0.0.1] - 3 Nov 2020

* beta version


test/animated_fab_menu_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:animated_fab_menu/animated_fab_menu.dart';

void main() {
  // test('adds one to input values', () {
  //   final calculator = Calculator();
  //   expect(calculator.addOne(2), 3);
  //   expect(calculator.addOne(-7), -6);
  //   expect(calculator.addOne(0), 1);
  //   expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  // });
}


lib/animated_fab_menu.dart
library animated_fab_menu;

import 'package:flutter/material.dart';

// ignore: must_be_immutable
class AddFab extends StatefulWidget {
  AddFab(
      {Key key,
      this.menu,
      this.btnSize = 50,
      this.expandHeight = 300,
      this.expandWidth = 200,
      this.color = Colors.blue})
      : super(key: key);

  Widget menu;
  double btnSize;
  double expandHeight;
  double expandWidth;
  Color color;

  @override
  _AddFabState createState() => _AddFabState();
}

class _AddFabState extends State<AddFab> {
  var _isMenuOpen = false;

  double _btnSize;
  double _height;
  double _width;
  Color _color;

  @override
  Widget build(BuildContext context) {
    _btnSize = widget.btnSize;
    _height = widget.expandHeight;
    _width = widget.expandWidth;
    _color = widget.color;

    return Container(
      width: double.infinity,
      height: double.infinity,
      child: Stack(
        alignment: Alignment.bottomRight,
        children: [
          GestureDetector(
            onTap: () => setState(() {
              this._isMenuOpen = false;
            }),
            child: AnimatedContainer(
              duration: Duration(milliseconds: 500),
              color: _isMenuOpen ? Colors.black54 : Colors.transparent,
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(20.0),
            child: AnimatedContainer(
              width: _isMenuOpen ? _width : _btnSize,
              height: _isMenuOpen ? _height : _btnSize,
              decoration: BoxDecoration(
                  shape: BoxShape.circle,
                  color: _isMenuOpen ? Colors.white : _color),
              alignment: AlignmentDirectional.bottomEnd,
              duration: Duration(milliseconds: 500),
              child: ConstrainedBox(
                  constraints: BoxConstraints(
                      maxHeight: _height,
                      maxWidth: _width,
                      minHeight: _btnSize,
                      minWidth: _btnSize),
                  child: Stack(
                    alignment: Alignment.center,
                    children: [
                      _isMenuOpen ? widget.menu : Container(),
                      _isMenuOpen
                          ? Container()
                          : GestureDetector(
                              onTap: () => setState(() {
                                this._isMenuOpen = true;
                              }),
                              child: Icon(
                                Icons.add,
                                color: Colors.white,
                                size: 30,
                              ),
                            ),
                    ],
                  )),
            ),
          )
        ],
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'animated_fab_menu.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('TestView'),
        centerTitle: true,
      ),
      body: SafeArea(
          child: Stack(
        children: [
          Center(
            child: Text('Hello World!'),
          ),
          AddFab(
            color: Colors.red,
            colorTo: Colors.white,
            expandHeight: 400,
            expandWidth: 150,
            btnSize: 70,
            icon: Icon(
              Icons.add,
              color: Colors.white,
              size: 40,
            ),
            menu: Container(
              padding: EdgeInsets.symmetric(vertical: 20),
              decoration: BoxDecoration(
                color: Colors.white,
                borderRadius: BorderRadius.circular(20),
              ),
              child: ListView.builder(
                itemCount: 6,
                itemBuilder: (context, i) => Container(
                  padding: EdgeInsets.symmetric(vertical: 5),
                  alignment: Alignment.center,
                  width: double.infinity,
                  decoration: BoxDecoration(color: Colors.white),
                  child: Text(
                    '$i',
                    style: TextStyle(fontSize: 25),
                  ),
                ),
              ),
            ),
          ),
        ],
      )),
    );
  }
}
